### PR TITLE
Update GetEiMonDet to version 3

### LINKED ISSUE #22833: GetEiMonDet should not rely on 'nominal' incident energy
The wavelength information in IN4 NeXus is sometimes way off. This prevents the usage of `GetEiMonDet` as the algorithm currently relies on the 'nominal' incident energy. This could be fixed by using the total TOF axis 'width' as a suitable lower limit for the TOF between the monitor and the detectors.

This change would allow `GetEiMonDet` to be used to not only calibrate, but unambiguously resolve Ei for IN4 and IN6.

**Original reporter:** Björn Fåk, ILL.

### PR DESCRIPTION
**Pull request by soininen**
**Description of work.**

This PR removes the heuristics relying on the nominal incident energy (as determined by the instrument's electronics/control software) and the elastic peak position tables which were used in `GetEiMonDet-v2`. Since the functionality and the algorithm's properties were changed, the version number was bumped to 3. Version 2 was marked as deprecated.

**Report to:** Björn Fåk, ILL 

**To test:**

Download the test file: 
[data.zip](https://github.com/mantidproject/mantid/files/2351597/data.zip)

The sample logs in the test file say the wavelength is 1.5Å. However, the instrument scientists say it should be around 1.3Å. For some reason the instrument control software has written a wrong value in the NeXus file.

Run the following script (fixing the path to the file):
```python
ws = LoadILLTOF('000001.nxs')
print('wavelength from sample logs: {}'.format(ws.run().getProperty('wavelength').value))
epps = FindEPP(ws)
pulseInterval = ws.run().getProperty('pulse_interval').value
E_i2 = GetEiMonDet(DetectorWorkspace=ws, IndexType='Workspace Index', 
    Detectors='0-299', DetectorEPPTable=epps, Monitor=396, PulseInterval=pulseInterval,
    Version=2)
lambda2 = UnitConversion.run('Energy', 'Wavelength', E_i2, ws.spectrumInfo().l1(), 2.0, 0., DeltaEModeType.Direct, 0)
print('wavelength from algorithm version 2: {}'.format(lambda2))
E_i3 = GetEiMonDet(DetectorWorkspace=ws, DetectorWorkspaceIndexSet='0-299',
    MonitorIndex=396)
lambda3 = UnitConversion.run('Energy', 'Wavelength', E_i3, ws.spectrumInfo().l1(), 2.0, 0., DeltaEModeType.Direct, 0)
print('wavelength from algorithm version 3: {}'.format(lambda3))
```

Observe, that version 2 of `GetEiMonDet` is fooled by the sample logs to calibrate the incident energy to corresponding wavelength 1.5Å while version 3 produces the expected result.

Fixes #22833.

---

#### Reviewer ####

Please comment on the following ([full description](http://developer.mantidproject.org/IndividualTicketTesting/)):

##### Code Review #####

- Is the code of an acceptable quality?
- Does the code conform to the [coding standards](http://developer.mantidproject.org/Standards/)?
- Are the unit tests small and test the class in isolation?
- If there are changes in the release notes then do they describe the changes appropriately?

##### Functional Tests #####

- Do changes function as described? Add comments below that describe the tests performed?
- Do the changes handle unexpected situations, e.g. bad input?
- Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.

### CODE REVIEW

File: Framework/Algorithms/src/GetEiMonDet3.cpp [hunk 1]
```diff
@@ -0,0 +1,305 @@
...
+#include "MantidKernel/EnabledWhenProperty.h"
+#include "MantidKernel/ListValidator.h"
+#include "MantidKernel/MandatoryValidator.h"
+#include "MantidKernel/PhysicalConstants.h"
+#include "MantidKernel/UnitConversion.h"
+#include "MantidKernel/VectorHelper.h"
+
+namespace {
+/** A private namespace holding the property names of
+ *  GetEiMonDet algorithm, version 2.
```
Comment (Mantid-Matthew) [inline Framework/Algorithms/src/GetEiMonDet3.cpp]: Should this read version 3 here?

File: Framework/Algorithms/src/GetEiMonDet3.cpp [hunk 2]
```diff
@@ -0,0 +1,305 @@
...
+/** Computes the minimum TOF between monitor and detectors from maximum
+ *  energy
+ *
+ * @param ws a workspace containing the instrument
+ * @param sampleToDetectorDistance the l2 distance
+ * @return minimum expected TOF, in microseconds
+ */
+double GetEiMonDet3::minimumTOF(const API::MatrixWorkspace &ws,
+                                const double sampleToDetectorDistance) {
+  const double minEnergy = getProperty(Prop::MAX_ENERGY);
```
Comment (Mantid-Matthew) [inline Framework/Algorithms/src/GetEiMonDet3.cpp]: would this make more sense being called maxEnergy or am I missing something?

**Review by Matthew-Andrew — COMMENTED** [review]
This looks good to me and the test script works as described. 

I have a few very minor comments below but otherwise all looks good.

File: Framework/Algorithms/src/GetEiMonDet3.cpp (hunk 2 above)
Comment (soininen) [reply Framework/Algorithms/src/GetEiMonDet3.cpp]: It should read `maxEnergy`, indeed. I think I got confused with the inverse proportionality of energy and TOF.

**Review by Matthew-Andrew — APPROVED** [review]
LGTM

**Comment by DanNixon** [comment]
This appears to have caused a significant performance regression in [AlgorithmsTest.SofQWNormalisedPolygonTestPerformance.testExec](http://builds.mantidproject.org/job/master_performancetests2/Master_20branch_20performance_20tests/AlgorithmsTest.SofQWNormalisedPolygonTestPerformance.testExec.htm).

**Comment by soininen** [comment]
@DanNixon Thanks for keeping track of the performance! I'll investigate the problem.

**Comment by soininen** [comment]
@DanNixon It appears that the performance test was completely rewritten in PR #24419 but somehow the tracks lead here. The test does a lot more work than before. There is a slight but known performance hit due to the new detector 2theta coverage computation but that is expected.

**Comment by DanNixon** [comment]
@soininen Ah, it looks like #24419 was merged just before this and only one performance tests job was scheduled.

**Pull request merged**